magma/syntax/combinational.py

```diff
@@ -210,69 +120,41 @@ def visit_Name(self, node):
-            ast.parse("from mantle import mux as phi").body[0],
-            tree
-        ])
     source = "\n"
     for i, line in enumerate(astor.to_source(tree).splitlines()):
         source += f"    {i}: {line}\n"
 
     debug(source)
-    circuit_def = ast_utils.compile_function_to_file(tree, fn.__name__,
-                                                     defn_env)
-    if get_debug_mode() and getattr(circuit_def, "debug_info", False):
-        circuit_def.debug_info = debug_info(circuit_def.debug_info.filename,
-                                            circuit_def.debug_info.lineno,
-                                            inspect.getmodule(fn))
+    return tree, env, metadata
 
-    @functools.wraps(fn)
-    def func(*args, **kwargs):
-        return circuit_def()(*args, **kwargs)
-    func.__name__ = fn.__name__
-    func.__qualname__ = fn.__name__
-    # Provide a mechanism for accessing the underlying circuit definition
-    setattr(func, "circuit_definition", circuit_def)
-    return func
 
-def combinational(
-        fn: typing.Callable = None,
-        *,
-        decorators: typing.Optional[typing.Sequence[typing.Callable]] = None,
-        ):
+class combinational(ast_tools.passes.Pass):
+    def rewrite(self, tree: ast.AST, env: ast_tools.SymbolTable,
+                metadata: dict):
+        tree, env, metadata = ast_tools.passes.ssa().rewrite(
+            tree, env, metadata)
 
-    exec(_SKIP_FRAME_DEBUG_STMT)
-    if decorators is not None:
-        assert fn is None
-        def wrapped(fn):
-            exec(_SKIP_FRAME_DEBUG_STMT)
-            nonlocal decorators
-            decorators = list(itertools.chain(decorators, [wrapped]))
-            wrapped_combinational = ast_utils.inspect_enclosing_env(
-                    _combinational,
-                    decorators=decorators)
-            return wrapped_combinational(fn)
-        return wrapped
+        phi_name = ast_tools.gen_free_name(tree, env,
+                                           "magma_combinational_phi")
 
-    else:
-        wrapped_combinational = ast_utils.inspect_enclosing_env(
-                _combinational,
-                decorators=[combinational]
-        )
-        return wrapped_combinational(fn)
+        if "mux" not in env:
+            from mantle import mux
+            mux_name = ast_tools.gen_free_name(tree, env,
+                                               "magma_combinational_mux")
+            env.locals[mux_name] = mux
+        else:
+            mux_name = "mux"
+
+        def phi(cond, arg0, arg1):
+            if all(isinstance(arg, tuple) for arg in (arg0, arg1)):
+                return tuple(phi(cond, *zipped) for zipped in zip(arg0,
+                                                                       arg1))
+            return env[mux_name]([arg1, arg0], cond)
+
+        env.locals[phi_name] = phi
+        tree, env, metadata = ast_tools.passes.if_to_phi(phi_name).rewrite(
```

Comment:
`if_to_phi` will generate a name for you with an optional prefix with the same logic that you have.  Just call:
```ast_tools.passes.if_to_phi(phi, 'magma_combinational_phi').rewrite(...)```

https://github.com/leonardt/ast_tools/blob/master/tests/test_if_to_phi.py#L23

https://github.com/leonardt/ast_tools/blob/master/ast_tools/passes/if_to_phi.py#L42